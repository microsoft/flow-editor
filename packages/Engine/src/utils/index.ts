export * from "./getPortPosition";
export * from "./getNodeConfig";
export * from "./graphDataUtils";
export * from "./transformMatrix";
export * from "./zoomAndPan";
export * from "./getCurvePathD";
export * from "./layout";
export * from "./browsers";
export * from "./isWithinThreshold";
export * from "./geometric";
export * from "./deepClone";
export * from "./diffGraphNode";
export * from "./diffGraphEdge";
export * from "./diffGraphPort";
export * from "./viewport";
export * from "./state";
export * from "./Writable";
export * from "./gantt/getCriticalPath";
export * from "./gantt/getTopoSortingNodes";
export * from "./getOffsetLimit";
export * from "./event";
export * from "./valueOrCallback";
export * from "./debug";
export * from "./debounce";
export * from "./arrayEquals";
export * from "./empty";
export * from "./stopPropagation";
export * from "./clamp";
export * from "./history";
