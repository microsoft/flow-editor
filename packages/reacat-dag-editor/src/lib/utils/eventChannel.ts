import { createRef } from "react";
import { IEvent } from "../models/event";
import { batchedUpdates } from "./batchedUpdates";

export type IGraphEventHandler<T = unknown, P = unknown, U = unknown> = (
  event: IEvent<T, P, U>
) => void;

export class EventChannel {
  public readonly listenersRef = createRef<IGraphEventHandler>();
  public readonly externalHandlerRef = createRef<
    IGraphEventHandler | undefined
  >();
  private queue: IEvent[] = [];
  private working = false;

  public trigger(event: IEvent): void {
    if (this.working) {
      this.queue.push(event);
    } else {
      this.working = true;
      batchedUpdates(() => {
        this.callHandlers(event);
        // eslint-disable-next-line @typescript-eslint/prefer-for-of
        for (let i = 0; i < this.queue.length; i += 1) {
          const e = this.queue[i];
          this.callHandlers(e);
        }
        this.queue = [];
      });
      this.working = false;
    }
  }

  public batch(events: IEvent[]): void {
    if (this.working) {
      this.queue.push(...events);
    } else {
      const first = events[0];
      if (!first) {
        return;
      }
      this.queue.push(...events.slice(1));
      this.trigger(first);
    }
  }

  private callHandlers(e: IEvent): void {
    this.listenersRef.current?.(e);
    if (!e.intercepted) {
      this.externalHandlerRef.current?.(e);
    }
  }
}
