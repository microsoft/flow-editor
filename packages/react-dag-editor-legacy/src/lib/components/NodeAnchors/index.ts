export * from "./INodeAnchors";
export * from "./GraphNodeAnchors";
export * from "./NodeAnchor";
