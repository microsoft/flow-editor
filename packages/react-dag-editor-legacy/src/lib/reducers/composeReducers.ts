import { IGraphReducer } from "../contexts";

export function composeReducers<
  NodeData = unknown,
  EdgeData = unknown,
  PortData = unknown,
  Action = never
>(
  reducers: ReadonlyArray<IGraphReducer<NodeData, EdgeData, PortData, Action>>
): IGraphReducer<NodeData, EdgeData, PortData, Action> {
  return (next) => reducers.reduceRight((prev, current) => current(prev), next);
}
