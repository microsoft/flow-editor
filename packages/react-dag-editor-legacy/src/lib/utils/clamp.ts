/**
 * @param min
 * @param max
 * @param value
 */
export function clamp(min: number, max: number, value: number): number {
  if (min > value) {
    return min;
  }
  if (max < value) {
    return max;
  }
  return value;
}
