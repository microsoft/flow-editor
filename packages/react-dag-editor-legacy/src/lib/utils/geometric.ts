import { IPoint, IRectShape } from "../models/geometry";

export const checkRectIntersect = (
  rectA: IRectShape,
  rectB: IRectShape
): boolean => {
  const isALeftOfB = rectA.maxX < rectB.minX;
  const isARightOfB = rectA.minX > rectB.maxX;
  const isABelowB = rectA.minY > rectB.maxY;
  const isAAboveB = rectA.maxY < rectB.minY;

  return !(isALeftOfB || isARightOfB || isABelowB || isAAboveB);
};

export const isPointInRect = (rect: IRectShape, point: IPoint): boolean => {
  const { minX, minY, maxX, maxY } = rect;
  const { x, y } = point;
  return x > minX && x < maxX && y > minY && y < maxY;
};

const square = (x: number): number => Math.pow(x, 2);

export const distance = (x1: number, y1: number, x2: number, y2: number) => {
  return Math.sqrt(square(x2 - x1) + square(y2 - y1));
};

export const getLinearFunction = (
  x1: number,
  y1: number,
  x2: number,
  y2: number
): ((x: number) => number) => {
  if (x1 === x2) {
    return () => Number.MAX_SAFE_INTEGER;
  }

  return (x: number) =>
    ((y2 - y1) / (x2 - x1)) * x + (y1 * x2 - y2 * x1) / (x2 - x1);
};

/**
 * (y1 - y2) * x + (x2 - x1) * y + (x1 * y2 - x2 * y1) = 0
 * a = y1 - y2
 * b = -(x2 - x1) = x1 - x2
 * c = -(x1 * y2 - x2 * y1) = x2 * y1 - x1 * y2
 *
 * a * x - b * y = c
 * x = (b * y + c) / a
 * y = (a * x - c) / b
 */
export class LinearEquation {
  private readonly a: number;
  private readonly b: number;
  private readonly c: number;

  public constructor(x1: number, y1: number, x2: number, y2: number) {
    this.a = y1 - y2;
    this.b = x1 - x2;
    this.c = x2 * y1 - x1 * y2;
  }

  public getX(y: number): number | undefined {
    if (this.a === 0) {
      return undefined;
    }
    return (this.b * y + this.c) / this.a;
  }

  public getY(x: number): number | undefined {
    if (this.b === 0) {
      return undefined;
    }
    return (this.a * x - this.c) / this.b;
  }
}
