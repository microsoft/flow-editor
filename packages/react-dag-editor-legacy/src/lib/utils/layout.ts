import type { OrderedMap } from "../collections";
import type { ICanvasGroup } from "../models/canvas";
import type { IGraphConfig, INodeConfig } from "../models/config/types";
import type { ICanvasNode } from "../models/node";
import type { NodeModel } from "../models/NodeModel";
import { getNodeConfig } from "./getNodeConfig";

/**
 * @param rectConfig
 * @param rect
 */
export function getRectWidth<T>(
  rectConfig: INodeConfig<T> | undefined,
  rect: Partial<ICanvasNode<T>>
): number {
  const minWidth = rectConfig?.getMinWidth(rect) ?? 0;

  if (rect.width && rect.width >= minWidth) {
    return rect.width;
  }

  return minWidth;
}

/**
 * @param rectConfig
 * @param rect
 */
export function getRectHeight<T>(
  rectConfig: INodeConfig<T> | undefined,
  rect: Partial<ICanvasNode<T>>
): number {
  const minHeight = rectConfig?.getMinHeight(rect) ?? 0;

  if (rect.height && rect.height >= minHeight) {
    return rect.height;
  }

  return minHeight;
}

export interface INodeRect {
  height: number;
  width: number;
}

/**
 * get node height and width by graphConfig
 *
 * @param node the node to get the size
 * @param graphConfig type IGraphConfig
 */
export function getNodeSize(
  node: ICanvasNode,
  graphConfig: IGraphConfig
): INodeRect {
  const nodeConfig = getNodeConfig(node, graphConfig);
  const width = getRectWidth(nodeConfig, node);
  const height = getRectHeight(nodeConfig, node);

  return {
    height,
    width,
  };
}

export interface IShapePosition {
  x: number;
  y: number;
}

export interface IShapeRect extends INodeRect, IShapePosition {}

/**
 * @param group
 * @param nodes
 * @param graphConfig
 */
export function getGroupRect<NodeData, PortData>(
  group: ICanvasGroup,
  nodes: OrderedMap<string, NodeModel<NodeData, PortData>>,
  graphConfig: IGraphConfig
): IShapeRect {
  const childrenIds = new Set(group.nodeIds);
  const children = Array.from(nodes.values()).filter((node) =>
    childrenIds.has(node.id)
  );
  const minX = Math.min(...children.map((node) => node.x));
  const maxX = Math.max(
    ...children.map((node) => node.x + getNodeSize(node, graphConfig).width)
  );
  const minY = Math.min(...children.map((node) => node.y));
  const maxY = Math.max(
    ...children.map((node) => node.y + getNodeSize(node, graphConfig).height)
  );

  const x = minX - (group.padding?.left ?? 0);
  const y = minY - (group.padding?.top ?? 0);
  const height = maxY - y + (group.padding?.bottom ?? 0);
  const width = maxX - x + (group.padding?.left ?? 0);

  return {
    x,
    y,
    width,
    height,
  };
}
