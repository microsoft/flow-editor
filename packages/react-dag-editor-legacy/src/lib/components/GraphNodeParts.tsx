import * as React from "react";
import { useVirtualization } from "../hooks/context";
import { useGetMouseDownOnAnchor } from "../hooks/useGetMouseDownOnAnchor";
import { NodeModel } from "../models/NodeModel";
import { isNodeEditing } from "../models/status";
import { isPointInRect } from "../utils";
import { GraphNode, IGraphNodeCommonProps } from "./GraphNode";
import { GraphNodeAnchors, RenderNodeAnchors } from "./NodeAnchors";
import {
  GraphOneNodePorts,
  IGraphOneNodePortsProps,
} from "./GraphOneNodePorts";

export interface IGraphNodePartsProps
  extends Required<Omit<IGraphNodeCommonProps, "viewport">>,
    Omit<IGraphOneNodePortsProps, "viewport"> {
  node: NodeModel;
  isNodeResizable: boolean;
  renderNodeAnchors?: RenderNodeAnchors;
}

const GraphNodeParts = ({
  node,
  isNodeResizable,
  renderNodeAnchors,
  ...commonProps
}: IGraphNodePartsProps) => {
  const virtualization = useVirtualization();
  const { renderedArea, viewport } = virtualization;
  const getMouseDown = useGetMouseDownOnAnchor(node, commonProps.eventChannel);
  const isVisible = isPointInRect(renderedArea, node);

  React.useLayoutEffect(() => {
    if (isVisible) {
      virtualization.renderedEdges.add(node.id);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [virtualization]);

  if (!isVisible) {
    return null;
  }

  let nodeAnchors;
  if (isNodeResizable && isNodeEditing(node)) {
    const defaultAnchors = (
      <GraphNodeAnchors node={node} getMouseDown={getMouseDown} />
    );
    nodeAnchors = renderNodeAnchors
      ? renderNodeAnchors(node, getMouseDown, defaultAnchors)
      : defaultAnchors;
  }

  return (
    <>
      <GraphNode {...commonProps} node={node} viewport={viewport} />
      <GraphOneNodePorts {...commonProps} node={node} viewport={viewport} />
      {nodeAnchors}
    </>
  );
};

const GraphNodePartsMemo = React.memo(GraphNodeParts);

export { GraphNodePartsMemo as GraphNodeParts };
