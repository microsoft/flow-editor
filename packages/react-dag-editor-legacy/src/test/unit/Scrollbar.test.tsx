import { render } from "@testing-library/react";
import * as React from "react";
import { EMPTY_GRAPH_STATE, IEvent, IGraphState, IViewport } from "../../index";
import { Scrollbar } from "../../lib/components/Scrollbar";
import { viewportReducer } from "../../lib/reducers/viewportReducer";
import { EventChannel } from "../../lib/utils/eventChannel";

describe("Scrollbar", () => {
  const ScrollbarComponent = () => {
    const rect: DOMRect | ClientRect = {
      width: 800,
      height: 600,
      left: 0,
      top: 0,
      right: 800,
      bottom: 800,
    };

    const [, dispatch] = React.useReducer(
      (prev: IGraphState, action: IEvent) => {
        return viewportReducer(prev, action);
      },
      {
        ...EMPTY_GRAPH_STATE,
      }
    );

    const viewport: Required<IViewport> = {
      rect,
      transformMatrix: [1, 0, 0, 1, 0, 0],
    };

    return (
      <Scrollbar
        viewport={viewport}
        offsetLimit={{ minX: -30, maxX: 20, minY: -100, maxY: 100 }}
        dispatch={dispatch}
        horizontal={true}
        vertical={true}
        eventChannel={new EventChannel()}
      />
    );
  };

  it("should match the snapshot", () => {
    const { container } = render(<ScrollbarComponent />);
    expect(container).toMatchSnapshot();
  });
});
