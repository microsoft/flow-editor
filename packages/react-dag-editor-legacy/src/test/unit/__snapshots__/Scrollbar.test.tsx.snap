// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrollbar should match the snapshot 1`] = `
<div>
  <div
    class="verticalScrollWrapper-0-2-1"
  >
    <div
      aria-label="vertical scrollbar"
      aria-roledescription="vertical scrollbar"
      class="verticalScrollStyle-0-2-3 verticalScrollStyle-d0-0-2-5"
      id="canvas-vertical-scrollbar"
      role="button"
    />
  </div>
  <div
    class="horizontalScrollWrapper-0-2-2"
  >
    <div
      aria-label="horizontal scrollbar"
      aria-roledescription="horizontal scrollbar"
      class="horizontalScrollStyle-0-2-4 horizontalScrollStyle-d1-0-2-6"
      id="canvas-horizontal-scrollbar"
      role="button"
    />
  </div>
</div>
`;
